cockatrice/src/client/ui/layouts/overlap_layout.cpp

```diff
@@ -0,0 +1,469 @@
+ *
+ * @return Integer count of items in the layout.
+ */
+int OverlapLayout::count() const
+{
+    return itemList.size();
+}
+
+/**
+ * @brief Provides access to a layout item at a specified index.
+ *
+ * Allows retrieval of a QLayoutItem from the layout’s internal list
+ * by index. If the index is out of bounds, this function will return nullptr.
+ *
+ * @param index The index of the desired item.
+ * @return Pointer to the QLayoutItem at the specified index, or nullptr if index is invalid.
+ */
+QLayoutItem *OverlapLayout::itemAt(int index) const
+{
+    return itemList.value(index);
+}
+
+/**
+ * @brief Removes and returns a layout item at the specified index.
+ *
+ * Removes a QLayoutItem from the layout at the given index, reducing the layout's count.
+ * If the index is invalid, this function returns nullptr without any effect.
+ *
+ * @param index The index of the item to remove.
+ * @return Pointer to the removed QLayoutItem, or nullptr if index is invalid.
+ */
+QLayoutItem *OverlapLayout::takeAt(int index)
+{
+    return itemList.takeAt(index);
+}
+
+/**
+ * @brief Sets the geometry for the layout items, arranging them with the specified overlap.
+ * @param rect The rectangle defining the area within which the layout should arrange items.
+ */
+void OverlapLayout::setGeometry(const QRect &rect)
+{
+    // Call the base class implementation to ensure standard layout behavior.
+    QLayout::setGeometry(rect);
+
+    // If there are no items to layout, exit early.
+    if (itemList.isEmpty()) {
+        return;
+    }
+
+    // Get the parent widget for size and margin calculations.
+    QWidget *parentWidget = this->parentWidget();
+    if (!parentWidget) {
+        return;
+    }
+
+    // Calculate available width and height, subtracting the parent's margins.
+    int availableWidth = parentWidget->width();
+    int availableHeight = parentWidget->height();
+    QMargins margins = parentWidget->contentsMargins();
+    availableWidth -= margins.left() + margins.right();
+    availableHeight -= margins.top() + margins.bottom();
+
+    // Determine the maximum item width and height among all layout items.
+    int maxItemWidth = 0;
+    int maxItemHeight = 0;
+    for (const QLayoutItem *item : itemList) {
+        if (item->widget()) {
+            QSize itemSize = item->widget()->sizeHint();
+            maxItemWidth = std::max(maxItemWidth, itemSize.width());
+            maxItemHeight = std::max(maxItemHeight, itemSize.height());
+        }
+    }
+
+    // Calculate the overlap offsets based on the layout direction and overlap percentage.
+    int overlapOffsetWidth = (direction == Qt::Horizontal) ? (maxItemWidth * overlapPercentage / 100) : 0;
+    int overlapOffsetHeight = (direction == Qt::Vertical) ? (maxItemHeight * overlapPercentage / 100) : 0;
+
+    // Determine the number of columns and rows based on layout constraints and available space.
+
+    int columns =
```

Comment:
I'm still not huge of the math here. Can we possibly split this up into using temporary, descriptive, variables?